bus-mapping/src/evm/opcodes/tstore.rs

```diff
@@ -0,0 +1,223 @@
+use super::Opcode;
+use crate::{
+    circuit_input_builder::{CircuitInputStateRef, ExecStep},
+    operation::{CallContextField, TransientStorageOp},
+    Error,
+};
+use eth_types::{GethExecStep, ToWord, Word};
+
+/// Placeholder structure used to implement [`Opcode`] trait over it
+/// corresponding to the [`OpcodeId::TSTORE`](crate::evm::OpcodeId::TSTORE)
+/// `OpcodeId`.
+#[derive(Debug, Copy, Clone)]
+pub(crate) struct Tstore;
+
+impl Opcode for Tstore {
+    fn gen_associated_ops(
+        state: &mut CircuitInputStateRef,
+        geth_steps: &[GethExecStep],
+    ) -> Result<Vec<ExecStep>, Error> {
+        let geth_step = &geth_steps[0];
+        let mut exec_step = state.new_step(geth_step)?;
+
+        let contract_addr = state.call()?.address;
+
+        state.call_context_read(
+            &mut exec_step,
+            state.call()?.call_id,
+            CallContextField::TxId,
+            Word::from(state.tx_ctx.id()),
+        )?;
+        state.call_context_read(
+            &mut exec_step,
+            state.call()?.call_id,
+            CallContextField::IsStatic,
+            Word::from(state.call()?.is_static as u8),
+        )?;
+
+        state.call_context_read(
+            &mut exec_step,
+            state.call()?.call_id,
+            CallContextField::RwCounterEndOfReversion,
+            Word::from(state.call()?.rw_counter_end_of_reversion),
+        )?;
+
+        state.call_context_read(
+            &mut exec_step,
+            state.call()?.call_id,
+            CallContextField::IsPersistent,
+            Word::from(state.call()?.is_persistent as u8),
+        )?;
+
+        state.call_context_read(
+            &mut exec_step,
+            state.call()?.call_id,
+            CallContextField::CalleeAddress,
+            state.call()?.address.to_word(),
+        )?;
+
+        // let key = geth_step.stack.nth_last(0)?;
+        // let key_stack_position = geth_step.stack.nth_last_filled(0);
+        // let value = geth_step.stack.nth_last(1)?;
+        // let value_stack_position = geth_step.stack.nth_last_filled(1);
+
+        let key = state.stack_pop(&mut exec_step)?;
+        let value = state.stack_pop(&mut exec_step)?;
+        #[cfg(feature = "enable-stack")]
+        {
+            assert_eq!(key, geth_step.stack.nth_last(0)?);
+            assert_eq!(value, geth_step.stack.nth_last(1)?);
+        }
+
+        // state.stack_read(&mut exec_step, key_stack_position, key)?;
+        // state.stack_read(&mut exec_step, value_stack_position, value)?;
+
+        let (_, value_prev) = state.sdb.get_transient_storage(&contract_addr, &key);
+        let value_prev = *value_prev;
```

Comment:
why we need value_prev?  (SSTORE needs value_prev since it is related to gas cost)

Reply:
oh revert.